src/torrent.cpp

```diff
@@ -12205,6 +12201,25 @@ namespace libtorrent
 		st->last_seen_complete = m_swarm_last_seen_complete;
 	}
 
+	void torrent::fill_pieces(bitfield* pieces, bool complete)
+	{
+		int num_pieces = m_torrent_file->num_pieces();
+		if (has_picker() && complete)
+		{
+			pieces->resize(num_pieces, false);
+			for (int i = 0; i < num_pieces; ++i)
+				if (m_picker->has_piece_passed(i)) pieces->set_bit(i);
```

Comment:
a piece having passed the hash check does not guarantee it has been flushed to disk. Even when we get a ``cache_flushed_alert``, we may have downloaded a new piece after the cache was flushed.

When a piece passes the hash check, ``async_flush_piece()`` is called to queue up the disk job to write it to disk. There's no explicit synchronization with this disk job completing and any piece picker or torrent state though.

Taking a step back. It seems like what you want to do is to query the disk cache of which pieces are dirty (i.e. not flushed to disk yet). You can do that with ``session_handle::get_cache_status()``. That's a synchronous call.